cprover_bindings/src/irep/goto_binary_serde.rs

```diff
@@ -0,0 +1,1339 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//! GOTO binary serializer.
+//! #![allow(dead_code, unused)]
+use memuse::DynamicUsage;
+
+use crate::irep::{Irep, IrepId, Symbol, SymbolTable};
+use crate::{InternString, InternedString};
+use std::collections::HashMap;
+use std::fs::File;
+use std::hash::Hash;
+use std::io::{self, BufReader};
+use std::io::{BufWriter, Bytes, Error, ErrorKind, Read, Write};
+use std::path::PathBuf;
+
+/// Writes a symbol table to a file in goto binary format in version 5.
+pub fn write_goto_binary_file(filename: &PathBuf, source: &crate::goto_program::SymbolTable) {
+    let out_file = File::create(filename).unwrap();
+    let mut writer = BufWriter::new(out_file);
+    let mut serializer = GotoBinarySerializer::new(&mut writer);
+    let irep_symbol_table = &source.to_irep();
+    serializer.write_file(irep_symbol_table).unwrap();
+}
+
+/// Reads a symbol table from a file expected to be in goto binary format in version 5.
+pub fn read_goto_binary_file(filename: &PathBuf) {
+    let file = File::open(filename).unwrap();
+    let reader = BufReader::new(file);
+    let mut deserializer = GotoBinaryDeserializer::new(reader);
+    deserializer.read_file().unwrap();
+}
+
+/// A numbered InternedString. The number is guaranteed to be in [0,N].
+/// Had to introduce this indirection because InternedString does not let you access
+/// its unique id, so we have to build one ourselves.
+#[derive(PartialEq, Eq, Hash, Clone, Copy, Debug)]
+struct NumberedString {
+    number: usize,
+    string: InternedString,
+}
+
+/// A key representing an Irep as the vector of unique numbers describing its contents.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+struct IrepKey {
+    numbers: Vec<usize>,
+}
+
+impl IrepKey {
+    /// The key is built as follows:
+    /// ```
+    /// id
+    /// sub.len()
+    /// sub[0]
+    /// ...
+    /// sub[sub.len()-1]
+    /// named_sub.len()
+    /// named_sub[0].0
+    /// named_sub[0].1
+    /// ...
+    /// named_sub[named_sub.len()-1].0
+    /// named_sub[named_sub.len()-1].1
+    /// ```
+    fn new(id: usize, sub: &[usize], named_sub: &[(usize, usize)]) -> Self {
+        let mut vec: Vec<usize> = Vec::new();
+        let size = sub.len() + 2 * named_sub.len() + 3;
+        vec.reserve_exact(size);
+        vec.push(id);
+        vec.push(sub.len());
+        vec.extend_from_slice(sub);
+        vec.push(named_sub.len());
+        for (k, v) in named_sub {
+            vec.push(*k);
+            vec.push(*v);
+        }
+        IrepKey { numbers: vec }
+    }
+}
+
+/// Inverse cache of unique NumberedIrep objects.
+struct IrepNumberingInv {
+    // Maps irep numbers to numbered ireps
+    index: Vec<NumberedIrep>,
+
+    // Stores the concactenation of all irep keys seen by the IrepNumbering owning this object
+    keys: Vec<usize>,
```

Comment:
Is this linearized (as opposed to using a two-dimensional vector) for performance reasons?

Reply:
This is a single vector storing the concatenation of all variable-length keys

Reply:
The goal with this is to not have to deal with references and lifetimes at all. Using Vec<IrepKey> and returning references to individual keys would reintroduce liftetimes everywhere.